File: app/src/main/java/com/timecapsule/app/data/repository/TimeCapsuleRepositoryImpl.kt
package com.timecapsule.app.data.repository

import com.timecapsule.app.data.local.dao.TimeCapsuleDao
import com.timecapsule.app.data.local.entity.TimeCapsuleEntity
import com.timecapsule.app.domain.model.TimeCapsule
import com.timecapsule.app.domain.repository.TimeCapsuleRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.time.LocalDateTime
import java.time.ZoneOffset

/**
 * Implementation of TimeCapsuleRepository using Room database.
 */
class TimeCapsuleRepositoryImpl(
    private val timeCapsuleDao: TimeCapsuleDao
) : TimeCapsuleRepository {

    override fun getAllCapsules(): Flow<List<TimeCapsule>> {
        return timeCapsuleDao.getAllCapsules().map { entities ->
            entities.map { it.toDomainModel() }
        }
    }

    override suspend fun getCapsuleById(id: Long): TimeCapsule? {
        return timeCapsuleDao.getCapsuleById(id)?.toDomainModel()
    }

    override suspend fun getUnopenedCapsules(): List<TimeCapsule> {
        return timeCapsuleDao.getUnopenedCapsules().map { it.toDomainModel() }
    }

    override suspend fun getUnlockedButNotNotifiedCapsules(): List<TimeCapsule> {
        val currentTime = LocalDateTime.now().toEpochSecond(ZoneOffset.UTC) * 1000
        return timeCapsuleDao.getUnlockedButNotNotifiedCapsules(currentTime).map { it.toDomainModel() }
    }

    override suspend fun createCapsule(timeCapsule: TimeCapsule): Long {
        return timeCapsuleDao.insertCapsule(TimeCapsuleEntity.fromDomainModel(timeCapsule))
    }

    override suspend fun updateCapsule(timeCapsule: TimeCapsule) {
        timeCapsuleDao.updateCapsule(TimeCapsuleEntity.fromDomainModel(timeCapsule))
    }

    override suspend fun deleteCapsule(id: Long) {
        timeCapsuleDao.deleteCapsuleById(id)
    }

    override suspend fun markAsOpened(id: Long) {
        timeCapsuleDao.markAsOpened(id)
    }

    override suspend fun markNotificationScheduled(id: Long) {
        timeCapsuleDao.markNotificationScheduled(id)
    }
}

File: lib/domain/repository/time_capsule_repository_impl.dart
import 'package:time_capsule/data/local/time_capsule_box.dart';
import 'package:time_capsule/domain/model/time_capsule.dart';
import 'package:time_capsule/domain/repository/time_capsule_repository.dart';

class TimeCapsuleRepositoryImpl implements TimeCapsuleRepository {
  final TimeCapsuleBox _box;

  TimeCapsuleRepositoryImpl(this._box);

  @override
  Future<List<TimeCapsule>> getAllCapsules() async {
    final capsules = _box.getAll().values.toList();
    capsules.sort((a, b) => b.createdAt.compareTo(a.createdAt));
    return capsules;
  }

  @override
  Future<TimeCapsule?> getCapsuleById(String id) async {
    return _box.get(id);
  }

  @override
  Future<List<TimeCapsule>> getUnopenedCapsules() async {
    final capsules = _box.getAll().values.toList();
    return capsules.where((c) => !c.isOpened).toList();
  }

  @override
  Future<List<TimeCapsule>> getUnlockedButNotNotifiedCapsules() async {
    final now = DateTime.now();
    final capsules = _box.getAll().values.toList();
    return capsules
        .where((c) => !c.isOpened && c.isUnlocked && !c.notificationScheduled)
        .toList();
  }

  @override
  Future<String> createCapsule(TimeCapsule capsule) async {
    await _box.put(capsule.id, capsule);
    return capsule.id;
  }

  @override
  Future<void> updateCapsule(TimeCapsule capsule) async {
    await _box.put(capsule.id, capsule);
  }

  @override
  Future<void> deleteCapsule(String id) async {
    await _box.delete(id);
  }

  @override
  Future<void> markAsOpened(String id) async {
    final capsule = await getCapsuleById(id);
    if (capsule != null && capsule.isUnlocked && !capsule.isOpened) {
      capsule.isOpened = true;
      await updateCapsule(capsule);
    }
  }

  @override
  Future<void> markNotificationScheduled(String id) async {
    final capsule = await getCapsuleById(id);
    if (capsule != null) {
      capsule.notificationScheduled = true;
      await updateCapsule(capsule);
    }
  }
}
